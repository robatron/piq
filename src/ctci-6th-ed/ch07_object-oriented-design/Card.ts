/**
 * Class and utilities representing a card of a standard 52 card deck. See
 * https://en.wikipedia.org/wiki/Standard_52-card_deck
 */

// Represents a card of a standard 52-card deck
export default class Card {
    private _rank: CardRank;
    private _suit: CardSuit;

    // Create a new Card from a card code or from a rank and suit
    constructor(val: CardRank | string, suit?: CardSuit) {
        // If suit is not defined, assume we have a card code, so attempt to
        // parse out the rank and suit from it
        if (suit === undefined) {
            this._rank = CardRankFromCode[val[0]];
            this._suit = CardSuitFromCode[val[1]];
        }

        // Otherwise, assume we just have a rank and suit
        else {
            this._rank = val as CardRank;
            this._suit = suit as CardSuit;
        }

        // Validate rank and suit, throw if either are invalid
        const isValidRank = Object.values(CardRank).includes(this._rank);
        const isValidSuit = Object.values(CardSuit).includes(this._suit);

        if (!isValidRank || !isValidSuit) {
            throw new Error(
                `Card ${
                    (!isValidRank ? `rank "${this._rank}"` : '') +
                    (!isValidSuit && !isValidRank ? ' and ' : '') +
                    (!isValidSuit ? `suit "${this._suit}"` : '')
                } is invalid (for val = ${val}, suit = ${suit})`,
            );
        }
    }

    get rank(): CardRank {
        return this._rank;
    }

    get suit(): CardSuit {
        return this._suit;
    }

    get code(): string {
        return `${CardRankToCode[this.rank]}${CardSuitToCode[this.suit]}`;
    }

    get name(): string {
        return `${this.rank} of ${this.suit}`;
    }

    get color(): CardColor {
        if (BLACK_SUITS.includes(this._suit)) {
            return CardColor.black;
        }
        return CardColor.red;
    }

    get isFaceCard(): boolean {
        return FACE_CARDS.includes(this._rank);
    }

    get isHonorCard(): boolean {
        return HONOR_CARDS.includes(this._rank);
    }

    get isNumberCard(): boolean {
        return NUMBER_CARDS.includes(this._rank);
    }
}

// Possible card colors
export enum CardColor {
    black = 'black',
    red = 'red',
}

// Card suits
export enum CardSuit {
    clubs = 'clubs',
    diamonds = 'diamonds',
    hearts = 'hearts',
    spades = 'spades',
}

// Card suit -> card suit code
export const CardSuitToCode = {
    [CardSuit.clubs]: 'C',
    [CardSuit.diamonds]: 'D',
    [CardSuit.hearts]: 'H',
    [CardSuit.spades]: 'S',
} as const;

// Card suit code -> card suit
export const CardSuitFromCode = {
    C: CardSuit.clubs,
    D: CardSuit.diamonds,
    H: CardSuit.hearts,
    S: CardSuit.spades,
} as const;

// Possible card ranks
export enum CardRank {
    ace = 'ace',
    two = 2,
    three = 3,
    four = 4,
    five = 5,
    six = 6,
    seven = 7,
    eight = 8,
    nine = 9,
    ten = 10,
    jack = 'jack',
    queen = 'queen',
    king = 'king',
}

// Card rank -> card rank code
export const CardRankToCode = {
    [CardRank.ace]: 'A',
    [CardRank.two]: '2',
    [CardRank.three]: '3',
    [CardRank.four]: '4',
    [CardRank.five]: '5',
    [CardRank.six]: '6',
    [CardRank.seven]: '7',
    [CardRank.eight]: '8',
    [CardRank.nine]: '9',
    [CardRank.ten]: 'T',
    [CardRank.jack]: 'J',
    [CardRank.queen]: 'Q',
    [CardRank.king]: 'K',
} as const;

// Card rank code -> card rank
export const CardRankFromCode = {
    A: CardRank.ace,
    '2': CardRank.two,
    '3': CardRank.three,
    '4': CardRank.four,
    '5': CardRank.five,
    '6': CardRank.six,
    '7': CardRank.seven,
    '8': CardRank.eight,
    '9': CardRank.nine,
    T: CardRank.ten,
    J: CardRank.jack,
    Q: CardRank.queen,
    K: CardRank.king,
} as const;

// All card suits
export const ALL_CARD_SUITS: ReadonlyArray<CardSuit> = [
    CardSuit.clubs,
    CardSuit.diamonds,
    CardSuit.hearts,
    CardSuit.spades,
];

// Just black suits
export const BLACK_SUITS: ReadonlyArray<CardSuit> = [
    CardSuit.clubs,
    CardSuit.spades,
];

// Just red suits (inverse black)
export const RED_SUITS: ReadonlyArray<CardSuit> = ALL_CARD_SUITS.filter(
    (cardSuit) => !BLACK_SUITS.includes(cardSuit),
);

// All card ranks, in order
export const ALL_CARD_RANKS: ReadonlyArray<CardRank> = [
    CardRank.ace,
    CardRank.two,
    CardRank.three,
    CardRank.four,
    CardRank.five,
    CardRank.six,
    CardRank.seven,
    CardRank.eight,
    CardRank.nine,
    CardRank.ten,
    CardRank.jack,
    CardRank.queen,
    CardRank.king,
];

// Face cards only
export const FACE_CARDS: ReadonlyArray<CardRank> = [
    CardRank.jack,
    CardRank.queen,
    CardRank.king,
];

// Honor cards (face cards + ace)
export const HONOR_CARDS: ReadonlyArray<CardRank> = [
    ...FACE_CARDS,
    CardRank.ace,
];

// Number cards (a.k.a. pip)
export const NUMBER_CARDS: ReadonlyArray<CardRank> = ALL_CARD_RANKS.filter(
    (cardRank) => !HONOR_CARDS.includes(cardRank),
);

// All 52 cards, in order
export const ALL_CARDS: ReadonlyArray<Card> = ((): Array<Card> => {
    const allCards: Array<Card> = [];

    ALL_CARD_SUITS.forEach((suit) => {
        ALL_CARD_RANKS.forEach((rank) => {
            allCards.push(new Card(rank, suit));
        });
    });

    return allCards;
})();

// ALL 52 card codes, in order
export const ALL_CARD_CODES: ReadonlyArray<string> = ((): Array<string> => {
    const allCardCodes: Array<string> = [];

    ALL_CARD_SUITS.forEach((suit) => {
        ALL_CARD_RANKS.forEach((rank) => {
            allCardCodes.push(`${CardRankToCode[rank]}${CardSuitToCode[suit]}`);
        });
    });

    return allCardCodes;
})();
