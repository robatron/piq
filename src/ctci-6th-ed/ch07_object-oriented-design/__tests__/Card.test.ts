import Card, {
    ALL_CARD_CODES,
    ALL_CARD_RANKS,
    ALL_CARD_SUITS,
    ALL_CARDS,
    BLACK_SUITS,
    CardColor,
    CardRank,
    CardRankFromCode,
    CardRankToCode,
    CardSuit,
    CardSuitToCode,
    FACE_CARDS,
    HONOR_CARDS,
    NUMBER_CARDS,
    RED_SUITS,
    CardSuitFromCode,
} from '../Card';

describe('Card', () => {
    describe('constructor', () => {
        it('can be instantiated with all possible cards', () => {
            ALL_CARD_SUITS.forEach((suit) => {
                ALL_CARD_RANKS.forEach((rank) => {
                    const card = new Card(rank, suit);
                    expect(card.rank).toBe(rank);
                    expect(card.suit).toBe(suit);
                });
            });
        });

        it('can be instantiated with all possible card codes', () => {
            ALL_CARD_SUITS.forEach((suit) => {
                ALL_CARD_RANKS.forEach((rank) => {
                    const card = new Card(
                        `${CardRankToCode[rank]}${CardSuitToCode[suit]}`,
                    );
                    expect(card.rank).toBe(rank);
                    expect(card.suit).toBe(suit);
                });
            });
        });

        describe('validation', () => {
            it('throws if the rank is invalid', () => {
                expect(() => {
                    new Card('invalid-rank' as CardRank, CardSuit.clubs);
                }).toThrowErrorMatchingInlineSnapshot(
                    `"Card rank \\"invalid-rank\\" is invalid (for val = invalid-rank, suit = clubs)"`,
                );

                expect(() => {
                    new Card('?C');
                }).toThrowErrorMatchingInlineSnapshot(
                    `"Card rank \\"undefined\\" is invalid (for val = ?C, suit = undefined)"`,
                );
            });

            it('throws if the suit is invalid', () => {
                expect(() => {
                    new Card(CardRank.ace, 'invalid-suit' as CardSuit);
                }).toThrowErrorMatchingInlineSnapshot(
                    `"Card suit \\"invalid-suit\\" is invalid (for val = ace, suit = invalid-suit)"`,
                );

                expect(() => {
                    new Card('A?');
                }).toThrowErrorMatchingInlineSnapshot(
                    `"Card suit \\"undefined\\" is invalid (for val = A?, suit = undefined)"`,
                );
            });

            it('throws if the rank and suit are invalid', () => {
                expect(() => {
                    new Card(
                        'invalid-rank' as CardRank,
                        'invalid-suit' as CardSuit,
                    );
                }).toThrowErrorMatchingInlineSnapshot(
                    `"Card rank \\"invalid-rank\\" and suit \\"invalid-suit\\" is invalid (for val = invalid-rank, suit = invalid-suit)"`,
                );

                expect(() => {
                    new Card('??');
                }).toThrowErrorMatchingInlineSnapshot(
                    `"Card rank \\"undefined\\" and suit \\"undefined\\" is invalid (for val = ??, suit = undefined)"`,
                );
            });
        });
    });

    it('reports its full name', () => {
        ALL_CARD_SUITS.forEach((cardSuit) => {
            ALL_CARD_RANKS.forEach((CardRank) => {
                const card = new Card(CardRank, cardSuit);
                expect(card.name).toBe(`${card.rank} of ${card.suit}`);
            });
        });
    });

    it('reports its short name', () => {
        ALL_CARD_SUITS.forEach((cardSuit) => {
            ALL_CARD_RANKS.forEach((CardRank) => {
                const card = new Card(CardRank, cardSuit);
                expect(card.code).toBe(
                    `${CardRankToCode[card.rank]}${CardSuitToCode[card.suit]}`,
                );
            });
        });
    });

    it('reports card color', () => {
        ALL_CARD_SUITS.forEach((suit) => {
            ALL_CARD_RANKS.forEach((rank) => {
                const card = new Card(rank, suit);
                if ([CardSuit.clubs, CardSuit.spades].includes(card.suit)) {
                    expect(card.color).toBe(CardColor.black);
                } else if (
                    [CardSuit.diamonds, CardSuit.hearts].includes(card.suit)
                ) {
                    expect(card.color).toBe(CardColor.red);
                } else {
                    throw new Error('Unexpected suit');
                }
            });
        });
    });

    it('reports if its a face card', () => {
        ALL_CARD_SUITS.forEach((suit) => {
            ALL_CARD_RANKS.forEach((rank) => {
                const card = new Card(rank, suit);
                if (
                    [CardRank.jack, CardRank.queen, CardRank.king].includes(
                        card.rank,
                    )
                ) {
                    expect(card.isFaceCard).toBe(true);
                } else {
                    expect(card.isFaceCard).toBe(false);
                }
            });
        });
    });

    it('reports if its an honor card', () => {
        ALL_CARD_SUITS.forEach((suit) => {
            ALL_CARD_RANKS.forEach((rank) => {
                const card = new Card(rank, suit);
                if (
                    [
                        CardRank.ace,
                        CardRank.jack,
                        CardRank.queen,
                        CardRank.king,
                    ].includes(card.rank)
                ) {
                    expect(card.isHonorCard).toBe(true);
                } else {
                    expect(card.isHonorCard).toBe(false);
                }
            });
        });
    });

    it('reports if its a number card', () => {
        ALL_CARD_SUITS.forEach((suit) => {
            ALL_CARD_RANKS.forEach((rank) => {
                const card = new Card(rank, suit);
                if (
                    ![
                        CardRank.ace,
                        CardRank.jack,
                        CardRank.queen,
                        CardRank.king,
                    ].includes(card.rank)
                ) {
                    expect(card.isNumberCard).toBe(true);
                } else {
                    expect(card.isNumberCard).toBe(false);
                }
            });
        });
    });
});

describe('module consts', () => {
    describe('CardColor', () => {
        it('should match expected values', () => {
            expect(CardColor).toMatchInlineSnapshot(`
                Object {
                  "black": "black",
                  "red": "red",
                }
            `);
        });
    });

    describe('CardSuit', () => {
        it('should match expected values', () => {
            expect(CardSuit).toMatchInlineSnapshot(`
                Object {
                  "clubs": "clubs",
                  "diamonds": "diamonds",
                  "hearts": "hearts",
                  "spades": "spades",
                }
            `);
        });
    });

    describe('CardSuitToCode', () => {
        it('should match expected values', () => {
            expect(CardSuitToCode).toMatchInlineSnapshot(`
                Object {
                  "clubs": "C",
                  "diamonds": "D",
                  "hearts": "H",
                  "spades": "S",
                }
            `);
        });
    });

    describe('CardSuitFromCode', () => {
        it('should match expected values', () => {
            expect(CardSuitFromCode).toMatchInlineSnapshot(`
                Object {
                  "C": "clubs",
                  "D": "diamonds",
                  "H": "hearts",
                  "S": "spades",
                }
            `);
        });
    });

    describe('CardRank', () => {
        it('should match expected values', () => {
            expect(CardRank).toMatchInlineSnapshot(`
                Object {
                  "10": "ten",
                  "2": "two",
                  "3": "three",
                  "4": "four",
                  "5": "five",
                  "6": "six",
                  "7": "seven",
                  "8": "eight",
                  "9": "nine",
                  "ace": "ace",
                  "eight": 8,
                  "five": 5,
                  "four": 4,
                  "jack": "jack",
                  "king": "king",
                  "nine": 9,
                  "queen": "queen",
                  "seven": 7,
                  "six": 6,
                  "ten": 10,
                  "three": 3,
                  "two": 2,
                }
            `);
        });
    });

    describe('CardRankToCode', () => {
        it('should match expected values', () => {
            expect(CardRankToCode).toMatchInlineSnapshot(`
                Object {
                  "10": "T",
                  "2": "2",
                  "3": "3",
                  "4": "4",
                  "5": "5",
                  "6": "6",
                  "7": "7",
                  "8": "8",
                  "9": "9",
                  "ace": "A",
                  "jack": "J",
                  "king": "K",
                  "queen": "Q",
                }
            `);
        });
    });

    describe('CardRankFromCode', () => {
        it('should match expected values', () => {
            expect(CardRankFromCode).toMatchInlineSnapshot(`
                Object {
                  "2": 2,
                  "3": 3,
                  "4": 4,
                  "5": 5,
                  "6": 6,
                  "7": 7,
                  "8": 8,
                  "9": 9,
                  "A": "ace",
                  "J": "jack",
                  "K": "king",
                  "Q": "queen",
                  "T": 10,
                }
            `);
        });
    });

    describe('ALL_CARD_SUITS', () => {
        it('should match expected values', () => {
            expect(ALL_CARD_SUITS).toMatchInlineSnapshot(`
                Array [
                  "clubs",
                  "diamonds",
                  "hearts",
                  "spades",
                ]
            `);
        });
    });

    describe('BLACK_SUITS', () => {
        it('should match expected values', () => {
            expect(BLACK_SUITS).toMatchInlineSnapshot(`
                Array [
                  "clubs",
                  "spades",
                ]
            `);
        });
    });

    describe('RED_SUITS', () => {
        it('should match expected values', () => {
            expect(RED_SUITS).toMatchInlineSnapshot(`
                Array [
                  "diamonds",
                  "hearts",
                ]
            `);
        });
    });

    describe('ALL_CARD_RANKS', () => {
        it('should match expected values', () => {
            expect(ALL_CARD_RANKS).toMatchInlineSnapshot(`
                Array [
                  "ace",
                  2,
                  3,
                  4,
                  5,
                  6,
                  7,
                  8,
                  9,
                  10,
                  "jack",
                  "queen",
                  "king",
                ]
            `);
        });
    });

    describe('FACE_CARDS', () => {
        it('should match expected values', () => {
            expect(FACE_CARDS).toMatchInlineSnapshot(`
                Array [
                  "jack",
                  "queen",
                  "king",
                ]
            `);
        });
    });

    describe('HONOR_CARDS', () => {
        it('should match expected values', () => {
            expect(HONOR_CARDS).toMatchInlineSnapshot(`
                Array [
                  "jack",
                  "queen",
                  "king",
                  "ace",
                ]
            `);
        });
    });

    describe('NUMBER_CARDS', () => {
        it('should match expected values', () => {
            expect(NUMBER_CARDS).toMatchInlineSnapshot(`
                Array [
                  2,
                  3,
                  4,
                  5,
                  6,
                  7,
                  8,
                  9,
                  10,
                ]
            `);
        });
    });

    describe('ALL_CARDS', () => {
        it('should contain all possible cards', () => {
            const allCards: Array<Card> = [];

            ALL_CARD_SUITS.forEach((suit) => {
                ALL_CARD_RANKS.forEach((rank) => {
                    allCards.push(new Card(rank, suit));
                });
            });

            expect(ALL_CARDS.length).toEqual(52);
            expect(ALL_CARDS).toStrictEqual(allCards);
        });

        it('should match expected values', () => {
            expect(ALL_CARDS).toMatchSnapshot();
        });
    });

    describe('ALL_CARD_CODES', () => {
        it('should contain all possible card codes', () => {
            const allCardCodes: Array<string> = [];

            ALL_CARD_SUITS.forEach((suit) => {
                ALL_CARD_RANKS.forEach((rank) => {
                    allCardCodes.push(
                        `${CardRankToCode[rank]}${CardSuitToCode[suit]}`,
                    );
                });
            });

            expect(ALL_CARD_CODES.length).toEqual(52);
            expect(ALL_CARD_CODES).toStrictEqual(allCardCodes);
        });

        it('should match expected values', () => {
            expect(ALL_CARD_CODES).toMatchSnapshot();
        });
    });
});
