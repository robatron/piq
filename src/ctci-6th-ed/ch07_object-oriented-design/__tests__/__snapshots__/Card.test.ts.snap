// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module consts ALL_CARD_CODES should match expected values 1`] = `
Array [
  "AC",
  "2C",
  "3C",
  "4C",
  "5C",
  "6C",
  "7C",
  "8C",
  "9C",
  "TC",
  "JC",
  "QC",
  "KC",
  "AD",
  "2D",
  "3D",
  "4D",
  "5D",
  "6D",
  "7D",
  "8D",
  "9D",
  "TD",
  "JD",
  "QD",
  "KD",
  "AH",
  "2H",
  "3H",
  "4H",
  "5H",
  "6H",
  "7H",
  "8H",
  "9H",
  "TH",
  "JH",
  "QH",
  "KH",
  "AS",
  "2S",
  "3S",
  "4S",
  "5S",
  "6S",
  "7S",
  "8S",
  "9S",
  "TS",
  "JS",
  "QS",
  "KS",
]
`;

exports[`module consts ALL_CARDS should match expected values 1`] = `
Array [
  Card {
    "_rank": "ace",
    "_suit": "clubs",
  },
  Card {
    "_rank": 2,
    "_suit": "clubs",
  },
  Card {
    "_rank": 3,
    "_suit": "clubs",
  },
  Card {
    "_rank": 4,
    "_suit": "clubs",
  },
  Card {
    "_rank": 5,
    "_suit": "clubs",
  },
  Card {
    "_rank": 6,
    "_suit": "clubs",
  },
  Card {
    "_rank": 7,
    "_suit": "clubs",
  },
  Card {
    "_rank": 8,
    "_suit": "clubs",
  },
  Card {
    "_rank": 9,
    "_suit": "clubs",
  },
  Card {
    "_rank": 10,
    "_suit": "clubs",
  },
  Card {
    "_rank": "jack",
    "_suit": "clubs",
  },
  Card {
    "_rank": "queen",
    "_suit": "clubs",
  },
  Card {
    "_rank": "king",
    "_suit": "clubs",
  },
  Card {
    "_rank": "ace",
    "_suit": "diamonds",
  },
  Card {
    "_rank": 2,
    "_suit": "diamonds",
  },
  Card {
    "_rank": 3,
    "_suit": "diamonds",
  },
  Card {
    "_rank": 4,
    "_suit": "diamonds",
  },
  Card {
    "_rank": 5,
    "_suit": "diamonds",
  },
  Card {
    "_rank": 6,
    "_suit": "diamonds",
  },
  Card {
    "_rank": 7,
    "_suit": "diamonds",
  },
  Card {
    "_rank": 8,
    "_suit": "diamonds",
  },
  Card {
    "_rank": 9,
    "_suit": "diamonds",
  },
  Card {
    "_rank": 10,
    "_suit": "diamonds",
  },
  Card {
    "_rank": "jack",
    "_suit": "diamonds",
  },
  Card {
    "_rank": "queen",
    "_suit": "diamonds",
  },
  Card {
    "_rank": "king",
    "_suit": "diamonds",
  },
  Card {
    "_rank": "ace",
    "_suit": "hearts",
  },
  Card {
    "_rank": 2,
    "_suit": "hearts",
  },
  Card {
    "_rank": 3,
    "_suit": "hearts",
  },
  Card {
    "_rank": 4,
    "_suit": "hearts",
  },
  Card {
    "_rank": 5,
    "_suit": "hearts",
  },
  Card {
    "_rank": 6,
    "_suit": "hearts",
  },
  Card {
    "_rank": 7,
    "_suit": "hearts",
  },
  Card {
    "_rank": 8,
    "_suit": "hearts",
  },
  Card {
    "_rank": 9,
    "_suit": "hearts",
  },
  Card {
    "_rank": 10,
    "_suit": "hearts",
  },
  Card {
    "_rank": "jack",
    "_suit": "hearts",
  },
  Card {
    "_rank": "queen",
    "_suit": "hearts",
  },
  Card {
    "_rank": "king",
    "_suit": "hearts",
  },
  Card {
    "_rank": "ace",
    "_suit": "spades",
  },
  Card {
    "_rank": 2,
    "_suit": "spades",
  },
  Card {
    "_rank": 3,
    "_suit": "spades",
  },
  Card {
    "_rank": 4,
    "_suit": "spades",
  },
  Card {
    "_rank": 5,
    "_suit": "spades",
  },
  Card {
    "_rank": 6,
    "_suit": "spades",
  },
  Card {
    "_rank": 7,
    "_suit": "spades",
  },
  Card {
    "_rank": 8,
    "_suit": "spades",
  },
  Card {
    "_rank": 9,
    "_suit": "spades",
  },
  Card {
    "_rank": 10,
    "_suit": "spades",
  },
  Card {
    "_rank": "jack",
    "_suit": "spades",
  },
  Card {
    "_rank": "queen",
    "_suit": "spades",
  },
  Card {
    "_rank": "king",
    "_suit": "spades",
  },
]
`;
